src/Compilers/CSharp/Portable/CodeGen/CodeGenerator_RefSafety.cs

```diff
@@ -0,0 +1,158 @@
+            parameters: method.Parameters,
+            arguments: node.Arguments,
+            argsToParamsOpt: default,
+            expanded: false);
+    }
+
+    private static bool MightEscapeTemporaryRefs(
+        bool used,
+        TypeSymbol returnType,
+        RefKind returnRefKind,
+        TypeSymbol? receiverType,
+        ScopedKind? receiverScope,
+        AddressKind? receiverAddressKind,
+        bool isReceiverReadOnly,
+        ImmutableArray<ParameterSymbol> parameters,
+        ImmutableArray<BoundExpression> arguments,
+        ImmutableArray<int> argsToParamsOpt,
+        bool expanded)
+    {
+        Debug.Assert(receiverAddressKind is null || receiverType is not null);
+
+        int writableRefs = 0;
+        int readonlyRefs = 0;
+
+        if (used && (returnRefKind != RefKind.None || returnType.IsRefLikeOrAllowsRefLikeType()))
+        {
+            writableRefs++;
+        }
+
+        if (receiverType is not null)
+        {
+            receiverScope ??= ScopedKind.None;
+            if (receiverAddressKind is { } a && !IsAnyReadOnly(a) && receiverScope == ScopedKind.None)
+            {
+                writableRefs++;
+            }
+            else if (receiverType.IsRefLikeOrAllowsRefLikeType() && receiverScope != ScopedKind.ScopedValue)
+            {
+                if (isReceiverReadOnly || receiverType.IsReadOnly)
+                {
+                    readonlyRefs++;
+                }
+                else
+                {
+                    writableRefs++;
+                }
+            }
+            else if (receiverAddressKind != null && receiverScope == ScopedKind.None)
+            {
+                readonlyRefs++;
+            }
+        }
+
+        if (shouldReturnTrue(writableRefs, readonlyRefs))
+        {
+            return true;
+        }
+
+        for (var arg = 0; arg < arguments.Length; arg++)
+        {
+            var parameter = Binder.GetCorrespondingParameter(
+                arg,
+                parameters,
+                argsToParamsOpt,
+                expanded);
+
+            if (parameter is not null)
+            {
+                if (parameter.RefKind.IsWritableReference() && parameter.EffectiveScope == ScopedKind.None)
+                {
+                    writableRefs++;
+                }
+                else if (parameter.Type.IsRefLikeOrAllowsRefLikeType() && parameter.EffectiveScope != ScopedKind.ScopedValue)
+                {
+                    if (parameter.Type.IsReadOnly || !parameter.RefKind.IsWritableReference())
+                    {
+                        readonlyRefs++;
+                    }
+                    else
+                    {
+                        writableRefs++;
+                    }
+                }
+                else if (parameter.RefKind != RefKind.None && parameter.EffectiveScope == ScopedKind.None)
+                {
+                    readonlyRefs++;
+                }
+            }
```

Comment:
>} [](http://example.com/codeflow?start=12&length=1)

What are the scenarios when we can fail to get a parameter and why they are not interesting?